pkg/virt-launcher/virtwrap/manager.go

```diff
@@ -641,6 +646,48 @@ func parseDeviceAddress(addrString string) []string {
 	return addrs
 }
 
+func expandDiskImagesOffline(vmi *v1.VirtualMachineInstance, domain *api.Domain) {
+	logger := log.Log.Object(vmi)
+	for _, disk := range domain.Spec.Devices.Disks {
+		if shouldExpand(disk) {
+			err := expandDiskImageOffline(disk.Source.File, disk.Size)
+			if err != nil {
+				logger.Reason(err).Errorf("failed to expand disk image %v at boot", disk)
+			}
+		}
+	}
+}
+
+func shouldExpand(disk api.Disk) bool {
+	if disk.AlwaysExpand {
+		return true
+	}
+	diskInfo, err := converter.GetImageInfo(getSourceFile(disk))
+	if err != nil {
+		return false
+	}
+	if diskInfo.VirtualSize >= disk.Size {
+		return false
+	}
+	return true
+}
+
+func expandDiskImageOffline(imagePath string, size int64) error {
+	log.Log.Infof("pre-start expansion of image %s to size %d", imagePath, size)
+	var preallocateFlag string
+	if converter.IsPreAllocated(imagePath) {
+		preallocateFlag = "--preallocation=falloc"
+	} else {
+		preallocateFlag = "--preallocation=off"
+	}
+	cmd := exec.Command("/usr/bin/qemu-img", "resize", preallocateFlag, imagePath, strconv.FormatInt(size, 10))
+	out, err := cmd.CombinedOutput()
```

Comment:
why is it safe for this function to assume the disk is a FS disk?

Reply:
It was previously safe because the argument would've been disk.Source.File which is "" for block devices.
I adjusted the code to avoid calling this function for block devices.